drivers: sensor: Add driver for TI INA23x

sambhurst opened the pull request:
This driver supports the TI INA230 and INA231 Bidirectional Current
and Power Monitors. The devices work on the I2C interface and are
created from DT nodes with a compatible property matching "ti,ina23x".

The following datasheets were referenced while developing the driver:
https://www.ti.com/product/INA230
https://www.ti.com/product/INA231

Testing was performed on the stm32g071b_disco board:
twister --west-flash --device-testing --device-serial /dev/ttyACM0 \
         -p stm32g071b_disco -T samples/sensor/ina23x

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

sambhurst committed:
drivers: sensor: Add driver for TI INA23x

This driver supports the TI INA230 and INA231 Bidirectional Current
and Power Monitors. The devices work on the I2C interface and are
created from DT nodes with a compatible property matching "ti,ina23x".

The following datasheets were referenced while developing the driver:
https://www.ti.com/product/INA230
https://www.ti.com/product/INA231

Twister passed:
twister -T tests/drivers/build_all/sensor/

Testing was performed on the stm32g071b_disco board with the following:
Load:    ~170 ohms
Voltage: 5V

Measured Values:
Voltage: 5.1 V
Current: 0.032 A
Power:   0.157 W

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

martinjaeger commented:
@carlescufi, @sambhurst I'm on vacation for about 2 weeks, but I'm happy to review this PR afterwards.

galak reviewed (changes requested):
Please add to the tests/drivers/build_all/sensor

galak reviewed (dismissed):
lets remove the sample as it doesn't do anything interesting

gmarull commented on drivers/sensor/ina23x/Kconfig:
```diff
@@ -0,0 +1,27 @@
+	  Set to enable trigger mode using gpio interrupt, where
+	  interrupts are configured to line ALERT PIN.
+
+config INA23X_CURRENT_LSB_1MA
+	bool "Set Current LSB to 1mA"
+	help
+	  When this option is enabled, the "current_lsb" device tree parameter
+	  is ignored and the Current is read in milliamps, Bus Voltage in
+	  millivolts, Shunt Voltage in microvolts and Power in milliwatts.
+	  When disabled, Current, Voltage, and Power registers are read in counts.
```
why can't this be a DT property?

sambhurst replied:
Done

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,360 @@
+#include "ina23x.h"
+
+/**
+ * @brife Macro for setting the current_lsb to 1mA
```
`@brief`

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,360 @@
+ * @param current_lsb  Value of the Current register LSB in milliamps
+ * @param rshunt       Shunt resistor value in milliohms
+ */
+ #define INA23X_CALIB(current_lsb, rshunt) (5120 / ((current_lsb) * (rshunt)))
```
extra space

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,360 @@
+/**
+ * @brief Macro to convert raw power value to milliwatts
+ *
+ * reg  value read from power register
```
`@param reg ...`

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,360 @@
+{
+	const struct ina23x_config *const config = dev->config;
+	struct ina23x_data *ina23x = dev->data;
+	uint8_t tx_buf[3] = { reg, (val >> 8), (val & 0xff) };
```
`byteorder.h` API should be used instead.

sambhurst replied:
Done

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,360 @@
+	uint8_t tx_buf[3] = { reg, (val >> 8), (val & 0xff) };
+
+	if (i2c_write(ina23x->i2c, tx_buf, sizeof(tx_buf), config->bus_addr) < 0) {
+		LOG_ERR("Failed to write register");
```
this log doesn't provide much information here (i2c driver logging can likely provide the same info)

gmarull commented on drivers/sensor/ina23x/ina23x.h:
```diff
@@ -0,0 +1,57 @@
+	uint16_t rshunt;
+	uint16_t mask;
+	uint16_t alert_limit;
+	const char *gpio_port;
+	gpio_pin_t irq_pin;
+	gpio_dt_flags_t irq_flags;
```
Better use `gpio_dt_spec`, see e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/lis2mdl/lis2mdl.c#L544

gmarull commented on drivers/sensor/ina23x/ina23x_trigger.c:
```diff
@@ -0,0 +1,66 @@
+	struct ina23x_data *ina23x = dev->data;
+	const struct ina23x_config *config = dev->config;
+
+	ina23x->gpio = device_get_binding(config->gpio_port);
```
when using `gpio_dt_spec`, just use `device_is_ready`, see https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/lis2mdl/lis2mdl_trigger.c#L127

gmarull commented on drivers/sensor/ina23x/ina23x_trigger.c:
```diff
@@ -0,0 +1,66 @@
+
+	ina23x->gpio = device_get_binding(config->gpio_port);
+	if (!ina23x->gpio) {
+		LOG_DBG("Gpio controller %s not found.", config->gpio_port);
```
`LOG_ERR`

gmarull reviewed (changes requested):
Thanks for this addition, please check comments

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,346 @@
+
+	switch (chan) {
+	case SENSOR_CHAN_BUS_VOLTAGE:
+		if (ina23x_reg_read(dev, INA23X_REG_BUS_VOLT,
+				    &ina23x->bus_voltage) < 0) {
+			LOG_ERR("Failed to read bus voltage");
+			return -EIO;
+		}
+		break;
+
+	case SENSOR_CHAN_CURRENT:
+		if (ina23x_reg_read(dev, INA23X_REG_CURRENT, &ina23x->current) < 0) {
+			LOG_ERR("Failed to read current");
+			return -EIO;
+		}
+		break;
+
+	case SENSOR_CHAN_POWER:
+		if (ina23x_reg_read(dev, INA23X_REG_POWER, &ina23x->power) < 0) {
+			LOG_ERR("Failed to read power");
+			return -EIO;
+		}
+		break;
+
+	case SENSOR_CHAN_SHUNT_VOLTAGE:
+		if (ina23x_reg_read(dev, INA23X_REG_SHUNT_VOLT,
+				    &ina23x->shunt_voltage) < 0) {
+			LOG_ERR("Failed to read shunt voltage");
+			return -EIO;
+		}
```
better propagate error code returned by `ina23x_reg_read`

sambhurst replied:
Done

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,346 @@
+
+	switch (attr) {
+	case SENSOR_ATTR_CONFIGURATION:
+		if (ina23x_reg_read(dev, INA23X_REG_CONFIG, &data) < 0) {
+			return -EIO;
+		}
+		break;
+	case SENSOR_ATTR_CALIBRATION:
+		if (ina23x_reg_read(dev, INA23X_REG_CALIB, &data) < 0) {
+			return -EIO;
+		}
+		break;
+	case SENSOR_ATTR_MASK:
+		if (ina23x_reg_read(dev, INA23X_REG_MASK, &data) < 0) {
+			return -EIO;
+		}
+		break;
+	case SENSOR_ATTR_ALERT:
+		if (ina23x_reg_read(dev, INA23X_REG_ALERT, &data) < 0) {
+			return -EIO;
+		}
+		break;
```
same comment here

sambhurst replied:
Done

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,346 @@
+		return -EINVAL;
+	}
+
+	if (ina23x_reg_write(dev, INA23X_REG_CONFIG, config->config) < 0) {
+		LOG_ERR("Failed to write configuration register!");
+		return -EINVAL;
+	}
+
+	cal = INA23X_CALIB(config->current_lsb, config->rshunt);
+
+	if (ina23x_reg_write(dev, INA23X_REG_CALIB, cal) < 0) {
+		LOG_ERR("Failed to write calibration register!");
+		return -EINVAL;
+	}
```
ditto

sambhurst replied:
Done

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,346 @@
+
+#ifdef CONFIG_INA23X_TRIGGER
+	if (config->trig_enabled) {
+		if (ina23x_trigger_mode_init(dev) < 0) {
+			LOG_ERR("Failed to init trigger mode\n");
+			return -EINVAL;
+		}
+
+		if (ina23x_reg_write(dev, INA23X_REG_ALERT,
+				     config->alert_limit) < 0) {
+			LOG_ERR("Failed to write alert register!");
+			return -EINVAL;
+		}
+
+		if (ina23x_reg_write(dev, INA23X_REG_MASK, config->mask) < 0) {
+			LOG_ERR("Failed to write mask register!");
+			return -EINVAL;
+		}
```
ditto

sambhurst replied:
Done

gmarull commented on include/dt-bindings/sensor/ina23x.h:
```diff
@@ -0,0 +1,73 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_INA23X_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_INA23X_H_
+
+/*
+ * Provide the BIT() macro for when this file is included from
+ * devicetrees.
+ */
+#ifndef BIT
+#define BIT(n) (1 << (n))
+#endif
```
would likely be better to make it available to DT

sambhurst replied:
Could you elaborate on this? I'm not sure what you mean.

gmarull replied:
sounds like a feature that would be useful to have on DT, so instead of providing a local solution, adjust dt utils.

sambhurst replied:
Got it. Thanks

mbolivar-nordic replied:
I think this is just a matter of moving BIT from util.h to util_macro.h. The latter is available in DT.

sambhurst replied:
Yes I agree. I'll move the BIT macros from util.h to util_macro.h. Should this be done in a separate pull request?

mbolivar-nordic replied:
I think it's OK to put it as the first commit in this PR as prep work personally

gmarull commented on dts/bindings/sensor/ti,ina23x.yaml:
```diff
@@ -0,0 +1,53 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: TI INA230 and INA231 Bidirectional Current and Power Monitor
```
can you document here how `include/dt-bindings/sensor/ina23x.h` is expected to be used?

gmarull commented on include/drivers/sensor.h:
```diff
@@ -294,6 +301,17 @@ enum sensor_attribute {
 	SENSOR_ATTR_CALIB_TARGET,
+	/** Configure the operating modes of a sensor. */
+	SENSOR_ATTR_CONFIGURATION,
+	/** */
```
missing docstring

gmarull commented on drivers/sensor/ina23x/ina23x_trigger.c:
```diff
@@ -0,0 +1,77 @@
+	/* setup alert gpio interrupt */
+	if (!device_is_ready(config->gpio_alert.port)) {
+		LOG_ERR("Cannot get pointer to alert_gpio device");
+		return -EINVAL;
```
```suggestion
		return -ENODEV;
```

sambhurst replied:
Done

gmarull commented on drivers/sensor/ina23x/ina23x_trigger.c:
```diff
@@ -0,0 +1,77 @@
+
+	/* setup alert gpio interrupt */
+	if (!device_is_ready(config->gpio_alert.port)) {
+		LOG_ERR("Cannot get pointer to alert_gpio device");
```
```suggestion
		LOG_ERR("Alert GPIO device not ready");
```

sambhurst replied:
Done

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,363 @@
+#define INA23X_DRIVER_INIT(inst)				    \
+	static struct ina23x_data drv_data_##inst;		    \
+	static const struct ina23x_config drv_config_##inst = {	    \
+		.bus_name = DT_INST_BUS_LABEL(inst),		    \
```
missed that one, better use `bus = DEVICE_DT_GET(DT_INST_BUS(inst))`, see e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/lis2mdl/lis2mdl.c#L593

sambhurst committed:
sys: util: Move BIT macros from util.h to util_macro.h

Moving the BIT macros from <sys/util.h> to <sys/util_macro.h>
allows the BIT macros to be used in device trees.

Testing: twister -T tests/drivers/build_all/sensor/

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

gmarull commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,353 @@
+	uint16_t cal;
+	int ret;
+
+	ret = ina23x_reg_write(dev, INA23X_REG_CONFIG, config->config);
```
Init function must check for bus device readyness, same as it is done for trigger gpio.

sambhurst replied:
Done

gmarull reviewed (approved):
LGTM

sambhurst commented:
@galak when you have time, could you take another look at this pull request? Thanks

martinjaeger commented on include/drivers/sensor.h:
```diff
@@ -173,6 +176,10 @@ enum sensor_channel {
 	SENSOR_CHAN_GAUGE_DESIRED_VOLTAGE,
 	/** Desired charging current in mA */
 	SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT,
+	/** Shunt voltage in uV */
+	SENSOR_CHAN_SHUNT_VOLTAGE,
+	/** Bus voltage in mV */
+	SENSOR_CHAN_BUS_VOLTAGE,
```
As stated above, shunt voltage is probably not needed.

Existing `SENSOR_CHAN_VOLTAGE` can be used instead of defining a new sensor channel.

sambhurst replied:
Done

martinjaeger commented on include/drivers/sensor.h:
```diff
@@ -294,6 +301,17 @@ enum sensor_attribute {
+	SENSOR_ATTR_CONFIGURATION,
+	/** Set a calibration value needed by a sensor. */
+	SENSOR_ATTR_CALIBRATION,
+	/**
+	 * Mask or Enable sensor functions usually related to
+	 * generating interrupts.
+	 */
+	SENSOR_ATTR_MASK,
+	/** Enable or Disable alerts and interrupts. */
+	SENSOR_ATTR_ALERT,
```
If I didn't understand wrong, the `ALERT` attribute sets the alert threshold. It is enabled in the `MASK` register. Mask is also a very generic term. I suggest the following change to make these attributes potentially more useful for other sensors:
```suggestion
	/** Enable/disable sensor features */
	SENSOR_ATTR_FEATURE_MASK,
	/** Alert threshold or alert enable/disable */
	SENSOR_ATTR_ALERT,
```

sambhurst replied:
Done

martinjaeger reviewed (commented):
As the sample was removed, the commit message should be adjusted.

sambhurst commented on drivers/sensor/ina23x/ina23x.c:
```diff
@@ -0,0 +1,358 @@
+	case SENSOR_CHAN_SHUNT_VOLTAGE:
+		val->val1 = INA2XX_SHUNT_UV(ina23x->shunt_voltage, config->current_lsb);
+		val->val2 = 0;
+		break;
```
Done

sambhurst commented on include/drivers/sensor.h:
```diff
@@ -121,6 +121,9 @@ enum sensor_channel {
 	/** Current, in amps **/
 	SENSOR_CHAN_CURRENT,
+	/** Power in mW */
+	SENSOR_CHAN_POWER,
```
I agree. This change uncovered some problems with the units.

sambhurst commented:
@galak I've made the changes you've requested. 
> Please add to the tests/drivers/build_all/sensor
> lets remove the sample as it doesn't do anything interesting

Are there any further changes you think I need to make? Thank you

mbolivar-nordic commented on drivers/sensor/ina23x/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_INA23X ina23x.c)
```
Redundant; see https://github.com/zephyrproject-rtos/zephyr/pull/36632

sambhurst replied:
Done

mbolivar-nordic commented on drivers/sensor/ina23x/ina23x_trigger.c:
```diff
@@ -0,0 +1,78 @@
+	ina23x->dev = dev;
+
+	ret = gpio_pin_configure_dt(&config->gpio_alert,
+				    GPIO_INPUT | config->gpio_alert.dt_flags);
```
Please remove ` | config->gpio_alert.dt_flags`; they are not needed with the `_dt` version of this routine.

sambhurst replied:
Done

mbolivar-nordic commented on dts/bindings/sensor/ti,ina23x.yaml:
```diff
@@ -0,0 +1,58 @@
+    current_lsb:
+      type: int
+      required: false
+      default: 1
+      # default of 1mA for current LSB
+      description: |
+        Value of Current LSB in milliamps. When set to 1mA
+        Current is read in mA, Bus Voltage in mV, Shunt
+        Voltage in uV, and Power in mW. Any other value
+        results in Current, Voltage, and Power registers
+        being read in counts.
```
You need to justify the use of `default` as described in https://docs.zephyrproject.org/latest/guides/dts/bindings.html#default-values-for-properties. E.g. if this is the power-on-reset value, then you must document this fact in the binding description.

sambhurst replied:
The default value was there to make life easier for the user but after ready the documentation, it seems better to let the user make a conscious decision about setting the value.

mbolivar-nordic commented on dts/bindings/sensor/ti,ina23x.yaml:
```diff
@@ -0,0 +1,58 @@
+    mask:
+      type: int
+      required: false
+      default: 0
+      # default all alert sources to disabled
```
The use of `default` must be justified in the description here also.

sambhurst replied:
Done

mbolivar-nordic commented on dts/bindings/sensor/ti,ina23x.yaml:
```diff
@@ -0,0 +1,58 @@
+    alert_limit:
+      type: int
+      required: false
+      default: 0
+      # default alert limit is 0V
```
The use of `default` must be justified in the description here also.

sambhurst replied:
Done

mbolivar-nordic reviewed (changes requested):
This review is limited mainly to the DT portions.

sambhurst commented:
@galak 
@mbolivar-nordic

I've made the changes you've requested. Are there any further changes you think I need to make? Thank you

mbolivar-nordic reviewed (approved):
Thanks. My comments are addressed.

sambhurst commented:
@galak
@nashif

Could you take another look at your earliest convenience? Thank you

cfriedt commented on include/sys/util.h:
```diff
@@ -305,46 +305,6 @@ uint8_t u8_to_dec(char *buf, uint8_t buflen, uint8_t value);
 /** @brief Number of Hz in @p x MHz */
 #define MHZ(x) (KHZ(x) * 1000)
 
-#ifndef BIT
```
Should this really be removed??
